akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -137,6 +143,91 @@ object GraphStageLogic {
   private object DoNothing extends (() ⇒ Unit) {
     def apply(): Unit = ()
   }
+
+  trait MinimalDeathWatch {
+    this: MinimalActorRef ⇒
+    private var watchedBy = ActorCell.emptyActorRefSet
+
+    def log: LoggingAdapter
+
+    //noinspection EmptyCheck
+    protected def sendTerminated(): Unit =
+      if (!watchedBy.isEmpty)
+        try {
+          def sendTerminated(ifLocal: Boolean)(watcher: ActorRef): Unit =
+            if (watcher.asInstanceOf[ActorRefScope].isLocal == ifLocal)
+              watcher.asInstanceOf[InternalActorRef].sendSystemMessage(DeathWatchNotification(this, existenceConfirmed = true, addressTerminated = false))
+
+          watchedBy foreach sendTerminated(ifLocal = false)
+          watchedBy foreach sendTerminated(ifLocal = true)
+        } finally watchedBy = ActorCell.emptyActorRefSet
+
+    override def stop(): Unit = sendTerminated()
+
+    protected def addWatcher(watchee: ActorRef, watcher: ActorRef): Unit = {
+      val watcheeSelf = watchee == this
+      val watcherSelf = watcher == this
+
+      if (watcheeSelf && !watcherSelf) {
+        if (!watchedBy.contains(watcher))
+          watchedBy += watcher
+      } else if (!watcheeSelf && watcherSelf) {
+        watch(watchee)
+      } else {
+        log.error("BUG: illegal Watch(%s,%s) for %s".format(watchee, watcher, this))
+      }
+    }
+
+    protected def remWatcher(watchee: ActorRef, watcher: ActorRef): Unit = {
+      val watcheeSelf = watchee == this
+      val watcherSelf = watcher == this
+
+      if (watcheeSelf && !watcherSelf) {
+        if (watchedBy.contains(watcher))
+          watchedBy -= watcher
+      } else if (!watcheeSelf && watcherSelf) {
+        unwatch(watchee)
+      } else {
+        log.error("BUG: illegal Unwatch(%s,%s) for %s".format(watchee, watcher, this))
+      }
+    }
+
+    def watch(actorRef: ActorRef): Unit =
+      actorRef.asInstanceOf[InternalActorRef].sendSystemMessage(Watch(actorRef.asInstanceOf[InternalActorRef], this))
+
+    def unwatch(actorRef: ActorRef): Unit =
+      actorRef.asInstanceOf[InternalActorRef].sendSystemMessage(Unwatch(actorRef.asInstanceOf[InternalActorRef], this))
+  }
+
+  /**
+   * Minimal actor to work with other actors and watch them in a synchronous ways
+   */
+  final class StageActorRef(val log: LoggingAdapter, asyncCallback: AsyncCallback[(ActorRef, Any)], override val path: ActorPath)
+    extends akka.actor.MinimalActorRef with MinimalDeathWatch {
+
+    override def provider: ActorRefProvider =
+      throw new UnsupportedOperationException("GraphStage MinimalActor does not provide")
```

Comment:
ok, cool. Will impl then, it's easy to fill in.
Thought there's something else this breaks